File: 0079-word-search/0079-word-search.cpp

int x[4] = {0, 0, 1, -1};
int y[4] = {1, -1, 0, 0};


class Solution {
public:
    int r, c;
    bool dfs(vector<vector<char>> &board, int a, int b, int pos, string word){
        if(board[a][b] != word[pos]) return false;
        if(pos == word.size() - 1) return true;
        board[a][b] = '$';
        bool res = false;
        for(int i = 0; i < 4; i++){
            int p = a + x[i];
            int q = b + y[i];
            if(p >= 0 && p < r && q >= 0 && q < c && res == false){
                res |= dfs(board, p, q, pos + 1, word);
            }
            if(res) return res;
        }
        board[a][b] = word[pos];
        return res;
    }
    bool exist(vector<vector<char>>& board, string word) {
        r = board.size();
        c = board[0].size();
        for(int i = 0; i < r; i++){
            for(int j = 0; j < c; j++){
                if(dfs(board, i, j, 0, word)) return true;
            }
        }
        return false;
    }
};

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def dfs(r, c, pos):
            if board[r][c] != word[pos]:
                return False
            if pos == len(word) - 1:
                return True
            board[r][c] = '$'
            res = False
            for i, j in [[0, 1], [0, -1], [1, 0], [-1, 0]]:
                nr = i + r
                nc = j + c
                if 0 <= nr < len(board) and 0 <= nc < len(board[0]) and not res:
                    res |= dfs(nr, nc, pos + 1)
                if res:
                    return res
            board[r][c] = word[pos]
            return res
        for r in range(len(board)):
            for c in range(len(board[0])):
                if dfs(r, c, 0):
                    return True
        return False
                